platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/exec/QueryModelRequirements.java

```diff
@@ -17,22 +17,25 @@
 package org.gradle.launcher.exec;
 
 import org.gradle.api.internal.StartParameterInternal;
+import org.gradle.internal.hash.HashCode;
 import org.gradle.internal.hash.Hasher;
 
+import java.util.function.Supplier;
+
 public class QueryModelRequirements extends AbstractToolingModelRequirements {
-    private final String modelName;
 
-    public QueryModelRequirements(StartParameterInternal startParameter,
-                                  boolean runsTasks,
-                                  String modelName) {
-        super(startParameter, runsTasks);
-        this.modelName = modelName;
+    public QueryModelRequirements(
+        StartParameterInternal startParameter,
+        boolean runsTasks,
+        Supplier<HashCode> payloadHashProvider
+    ) {
+        super(startParameter, runsTasks, payloadHashProvider);
     }
 
     @Override
     public void appendKeyTo(Hasher hasher) {
         // Identify the type of action
         hasher.putByte((byte) 2);
-        hasher.putString(modelName);
+        hasher.putHash(getPayloadHash());
```

Comment:
💅 It seems the whole `appendKeyTo` implementation could now be moved up and each subtype would implement something like `byte getActionId()`

Reply:
I lifted the hashing logic to the abstract class. 

Now the classes look almost exactly the same. Though, I think it's okay to preserve them as separate classes. They may still evolve differently in the future.